core/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java

```diff
@@ -518,7 +573,7 @@ protected void runReplicateTest(IndexShardRoutingTable shardRoutingTable, int as
             if (shard.unassigned()) {
                 continue;
             }
-            if (shard.primary() == false) {
+            if (shard.primary() == false || shard.relocating()) { // for relocating primaries, we replica from target to source if source is marked as relocated
                 nodesSentTo.remove(shard.currentNodeId());
             }
             if (shard.relocating()) {
```

Comment:
I'm not sure I understand this comment correctly. I pushed a new commit where I have added assertions: https://github.com/ywelsch/elasticsearch/commit/473464c6d4ac5b3837affdd1d7e596c966929951#diff-3988756cc764d148f763c9d0cb740c61R565